### PR TITLE
Use a single function for query manipulations

### PR DESCRIPTION
**Pull request by cjgillot**
There are three ways to invoke a query: get the value, ensure it runs, or have it automatically forced.

The three corresponding methods are merged into a single point of entry.
The goal is to increase the code sharing between the three cases.

### CODE REVIEW

**Comment by Xanewok** [comment]
I vaguely recall that this may have affected performance (e.g. see careful `#[inline(never)]` and `#[inline(always)]` annotations), let's see if this has any impact:

@bors try @rust-timer queue

File: compiler/rustc_query_system/src/query/plumbing.rs [hunk 1]
```diff
@@ -572,17 +575,15 @@ fn incremental_verify_ich<CTX, K, V>(
...
-    job: JobOwner<'_, CTX, C>,
+    key: K,
+    job_id: QueryJobId<CTX::DepKind>,
     dep_node: DepNode<CTX::DepKind>,
-    query: &QueryVtable<CTX, C::Key, C::Value>,
-) -> (C::Stored, DepNodeIndex)
+    query: &QueryVtable<CTX, K, V>,
+) -> (V, DepNodeIndex)
 where
-    C: QueryCache,
```
Comment (Julian-Wollersberger) [inline compiler/rustc_query_system/src/query/plumbing.rs:567]: Glad to see this `C: QueryCache` clause removed ^^
It is annoying when trying to make things `dyn`.

**Review by Julian-Wollersberger — CHANGES_REQUESTED** [review]
This refactoring should improve bootstrap times a bit, because it removes 216 000 [llvm-lines](https://rustc-dev-guide.rust-lang.org/profiling.html#optimizing-rustcs-bootstrap-times-with-cargo-llvm-lines) (-1%).

```
Before:  Lines   Copies         Function name
--------------   ------         -------------
20218434 (100%)  600060 (100%)  (TOTAL)
  457651 (2.3%)     414 (0.1%)  rustc_query_system::query::plumbing::get_query_impl::{{closure}}
  168207 (0.8%)     354 (0.1%)  rustc_query_system::query::plumbing::force_query_impl::{{closure}}
   95634 (0.5%)     207 (0.0%)  rustc_query_system::query::plumbing::incremental_verify_ich
   87688 (0.4%)    4335 (0.7%)  core::ops::function::FnOnce::call_once
   75830 (0.4%)     621 (0.1%)  <rustc_query_system::query::caches::DefaultCache<K,V> as rustc_query_system::query::caches::QueryCache>::iter
   69233 (0.3%)     207 (0.0%)  rustc_query_system::query::plumbing::load_from_disk_and_cache_in_memory
   41047 (0.2%)     207 (0.0%)  rustc_query_system::query::plumbing::get_query_impl

After:  Lines    Copies         Function name
-------------    ------         -------------
20001669 (100%)  595297 (100%)  (TOTAL)
  434107 (2.2%)     354 (0.1%)  rustc_query_system::query::plumbing::call_query_impl::{{closure}}
   85516 (0.4%)    4257 (0.7%)  core::ops::function::FnOnce::call_once
   81774 (0.4%)     177 (0.0%)  rustc_query_system::query::plumbing::incremental_verify_ich
   75830 (0.4%)     621 (0.1%)  <rustc_query_system::query::caches::DefaultCache<K,V> as rustc_query_system::query::caches::QueryCache>::iter
   59176 (0.3%)     177 (0.0%)  rustc_query_system::query::plumbing::load_from_disk_and_cache_in_memory
   47042 (0.2%)     177 (0.0%)  rustc_query_system::query::plumbing::call_query_impl
```
(Omitted all entries from other crates; mostly core, alloc and hashbrown.)

File: compiler/rustc_middle/src/ty/query/plumbing.rs [hunk 2]
```diff
@@ -394,6 +394,19 @@ macro_rules! define_queries_inner {
...
+        impl queries::$name<$tcx> {
+            $(#[$attr])*
+            #[inline(always)]
+            pub fn query(
+                tcx: TyCtxt<$tcx>,
+                span: Span,
+                key: query_keys::$name<$tcx>,
+                caller: QueryCaller<DepKind>,
+            ) -> Option<<queries::$name<$tcx> as QueryConfig<TyCtxt<$tcx>>>::Stored> {
+                call_query::<queries::$name<'_>, _>(tcx, span, key, caller)
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_middle/src/ty/query/plumbing.rs:408]: The point is to monomorphize `call_query` only once for each query. The easiest way is to call it through a monomorphic function.

**Comment by cjgillot** [comment]
LLVM-lines was the target, but perf is bad. Bootstrap times are not a win either. This definitely needs more work.

File: compiler/rustc_middle/src/ty/query/plumbing.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Julian-Wollersberger) [reply compiler/rustc_middle/src/ty/query/plumbing.rs:408]: I just tested removing the wrapper. It removed another 39 022 llvm-lines.
I guess `call_query` is only monomorphized once for each query anyway, even though it is called from multiple places.

**Comment by cjgillot** [comment]
Closed in favor of #78780.

**Pull request closed without merging**